stress/README.md

```diff
@@ -0,0 +1,36 @@
+# OpenTelemetry Stress Tests
+
+## Why would you need stress test
+
+* It helps you to understand performance.
+* You can keep it running for days and nights to verify stability.
+* You can use it to generate lots of load to your backend system.
+* You can use it with other stress tools (e.g. a memory limiter) to verify how
+  your code reacts to certain resource constraints.
+
+## Usage
+
+Open a console, run the following command from the current folder:
+
+```sh
+cargo run --release --bin X
+```
+
+where `X` is the specific stress test you would like to run.
+
+e.g.
+
+```sh
+cargo run --release --bin metrics
+```
+
+```text
+Throughput: 70878.00 requests/sec
```

Comment:
this is very low (we should be seeing millions/sec, as each call is only 100s of nanosecs from the cargo bench runs). I think the calculation or orchestration is not fully correct.

Reply:
I forgot to change the README but the execution is actually somewhere around 400k/sec

Reply:
I think it should be ~15 Million/sec, based on the ~65ns it takes to make s counter.Add(1) with no Attributes.

Reply:
I'm getting about 4 Million/Sec now with the new changes. I wonder if acquiring the lock to update the counts within the closure adds a bit of overhead.

Reply:
@jtescher @TommyCpp  @hdost 

In your experience, would obtaining a lock take that much time in `let mut counts = counts.lock().unwrap();`? If so, not sure how to implement this to allow multiple threads to modify the same vector. @cijothomas  if not possible, are we able to use the current readings as a baseline, even though it is "slower" than what we expect?

Reply:
Not sure, but will it help if we use message passing (std::sync::mpsc;), so each of the threads would write it's local count to message-queue every few ms (or lesser - maybe every increment), and main loop will keep receiving them in it's loop and adding. I assume read and writes to the queue are lockfree, so at-least it shouldn't have overhead for lock. Just a thought if this can be tried.

Reply:
One simple replacement is `AtomicU64` instead of a locked int

Reply:
> In your experience, would obtaining a lock take that much time in let mut counts = counts.lock().unwrap()

I believe that in this case, the counts variable will experience significant contention because multiple threads are attempting to lock it.

Reply:
Also second @lalitb we probably want to keep local variable local and throw them into queue. One choice can be https://crates.io/crates/crossbeam-queue

Reply:
@TommyCpp 
I'll give that a try.

Reply:
Attempt #1 with atomic u64:

```
Throughput: 4813019.00 requests/sec
Throughput: 4805997.50 requests/sec
Throughput: 4796934.50 requests/sec
Throughput: 4817836.50 requests/sec
Throughput: 4812457.00 requests/sec
Throughput: 4773890.50 requests/sec
Throughput: 4760047.00 requests/sec
Throughput: 4851061.50 requests/sec
```
Seems like incrementing an atomic u64 still has some overhead. Will continue experimenting...

Reply:
Is it possible that we're running into a limitation in our implementation? And that we need to optimize the SDK counter? I noticed that in the hot path that we're doing Aggregate, so I'm wondering if that's not something we should consider moving. I'd have assumed that just like with our stress tester we'd want to have the actual aggregation on the alternative thread.

Reply:
For existing [implementation](https://github.com/open-telemetry/opentelemetry-rust/pull/1045/commits/97b179c532bdf8c790d1a775d6221e54da9470d2) I get
```
Throughput: 813044.50 requests/sec
Throughput: 824154.50 requests/sec
Throughput: 827773.50 requests/sec
Throughput: 806773.50 requests/sec
Throughput: 820415.50 requests/sec
Throughput: 813496.00 requests/sec
Throughput: 813064.50 requests/sec
Throughput: 807269.50 requests/sec
Throughput: 815698.50 requests/sec
Throughput: 803485.50 requests/sec
Throughput: 799224.00 requests/sec
Throughput: 799852.00 requests/sec
Throughput: 789417.50 requests/sec
Throughput: 785463.50 requests/sec
```

I think we can future optimize by having a vector of atomic number reference. i.e 
```rust
  let mut total_count = vec![];
    let mut total_count_refs = vec![];
    for _ in 0..num_cpus::get() {
        let count = Arc::new(AtomicU64::new(0));
        total_count.push(count.clone());
        total_count_refs.push(count);
    }

    rayon::spawn( move || {
        total_count.into_par_iter().for_each(|count| loop {
            func();
            count.fetch_add(1, Ordering::SeqCst);
            if STOP.load(Ordering::SeqCst) {
                break;
            }
        });
    });

    loop {
        let elapsed = end_time.duration_since(start_time).as_secs();
        if elapsed >= SLIDING_WINDOW_SIZE {
            let mut total_count_u64 = 0;
            for i in 0..num_cpus::get(){
                total_count_u64 += total_count_refs[i].load(Ordering::Relaxed);
            }
            let current_count = total_count_u64 - total_count_old;
            total_count_old = total_count_u64;
            let throughput = current_count as f64 / elapsed as f64;
            println!("Throughput: {:.2} requests/sec", throughput);
            start_time = Instant::now();
        }

        if STOP.load(Ordering::SeqCst) {
            break;
        }

        end_time = Instant::now();
    }
```
I can get slightly better result 
```
Throughput: 917330.50 requests/sec
Throughput: 915197.00 requests/sec
Throughput: 930210.50 requests/sec
Throughput: 919314.00 requests/sec
Throughput: 932616.50 requests/sec
Throughput: 913703.00 requests/sec
```

Reply:
> Can someone else try to run this? It could possibly be my cpu.

I see slightly better performance with existing implementation in this PR, of-course this is the release build:
```
$ cargo run --release
    Finished release [optimized] target(s) in 0.11s
     Running `/tmp/tbd2/opentelemetry-rust/target/release/metrics`
Throughput: 6086848.00 requests/sec
Throughput: 6212395.50 requests/sec
Throughput: 6218707.00 requests/sec
Throughput: 6244260.00 requests/sec
Throughput: 6220992.00 requests/sec
Throughput: 5648886.50 requests/sec
Throughput: 6265541.50 requests/sec
Throughput: 5894563.50 requests/sec
Throughput: 6267127.50 requests/sec
Throughput: 5647298.00 requests/sec
```

Reply:
> @TommyCpp I get similar results using your code as with the current implementation with atomicu64:
> 
> ```
> Throughput: 4542732.50 requests/sec
> Throughput: 4549933.50 requests/sec
> Throughput: 4427950.50 requests/sec
> Throughput: 4525449.50 requests/sec
> Throughput: 4521940.50 requests/sec
> Throughput: 4538587.00 requests/sec
> ```
> 
> Could we maybe merge as is and then iterate on the solution? @cijothomas wdyt

Yeah I think we can merge as it is and keep iterating

Reply:
> @hdost We are comparing with the ~60ns it takes to run this function on average once as shown in the [microbenchmarking](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/benches/metric.rs#L85) results. By this, (even with our current "slow" implementation) we _should_ be getting around ~15m calls/second. Is there some nuances in Rust itself or the way we are setting up the test that is adding some additional overhead that doesn't translate to simply 2 seconds / 60ns = 15m ? @cijothomas wdyt

My guess is the microbenchmarking (cargo bench) does not actually run multiple threads, but this stress test does, so microbenchmarking is not showing lock contention in its measurements, but stress test will include contention. So the stress may show much lower than the 15m/sec we anticipate based on cargo bench results.